File: app/src/main/java/com/xn1ch1/qrscavengerhunt/ActivitySetClues.java
package com.xn1ch1.qrscavengerhunt;

import android.app.Activity;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.view.ContextThemeWrapper;
import android.util.TypedValue;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;

public class ActivitySetClues extends AppCompatActivity {

    SharedPreferences mDataStore;
    SharedPreferences.Editor mDataStoreEditor;

    InputMethodManager mInputMethod;

    int mRowID = 0;

    String[][] mClueData = new String[10][2];

    int[] mClueChecks = new int[2];
    int[] mCodeChecks = new int[2];
    float[] mCheckStatus = new float[2];

    TypedValue mOutValue = new TypedValue();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_set_clues);


        getResources().getValue(R.dimen.allViewSemiTransparent, mOutValue, true);
        mCheckStatus[0] = mOutValue.getFloat();
        getResources().getValue(R.dimen.allViewEnabled, mOutValue, true);
        mCheckStatus[1] = mOutValue.getFloat();

        mCodeChecks[0] = R.drawable.set_clue_scan_unchecked;
        mCodeChecks[1] = R.drawable.set_clue_scan_checked;

        mClueChecks[0] = R.drawable.set_clue_clue_text_unchecked;
        mClueChecks[1] = R.drawable.set_clue_clue_text_checked;

        loadClueData();
        setAllTextAndChecks();
        textInputInvisible();

    }

    @Override
    public void onBackPressed() {
        if (findViewById(R.id.SetCluesTextInputOverlay).getVisibility() == View.VISIBLE) {
            saveTextInput();
            setSingleTextAndChecks(mRowID);
            textInputInvisible();
        } else {
            super.onBackPressed();
            writeClueData();
            Intent intent = new Intent(this, ActivityHome.class);
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            startActivity(intent);
        }
    }

    @Override
    protected void onActivityResult(int mRequestCode, int mResultCode, Intent mData) {
        if (mResultCode == Activity.RESULT_OK) {

            String mReturnedCode = mData.getStringExtra("code");

            if (codeAlreadyAssigned(mReturnedCode)) {
                AlertDialog.Builder builder = new AlertDialog.Builder(new ContextThemeWrapper(this, R.style.AlertDialogCustom));
                builder.setMessage(getResources().getString(R.string.setCluesCodeUsed, mReturnedCode));
                builder.setNeutralButton(getResources().getString(R.string.ok), null).show();
            } else {
                mClueData[mRequestCode][1] = mReturnedCode;
                writeClueData();
                setSingleTextAndChecks(mRequestCode);

                final AlertDialog.Builder builder = new AlertDialog.Builder(new ContextThemeWrapper(this, R.style.AlertDialogCustom));
                builder.setMessage(getResources().getString(R.string.setCluesCodeOK, mReturnedCode, mRequestCode + 1));
                builder.setNeutralButton(getResources().getString(R.string.ok), null);
                builder.show();
            }
        }
    }

    /* ******************
        User Interaction
       ****************** */
    public void inputClueText(View view) {
        mRowID = view.getId();
        String mRowName = getResources().getResourceEntryName(mRowID);
        mRowName = mRowName.substring(mRowName.length() - 1);
        mRowID = Integer.parseInt(mRowName);
        textInputVisible(mRowID);
    }

    public void setCluesScanBarcode(View view) {

        mRowID = view.getId();
        String mRowName = getResources().getResourceEntryName(mRowID);
        mRowName = mRowName.substring(mRowName.length() - 1);
        mRowID = Integer.parseInt(mRowName);

        if (mClueData[mRowID][1].equals("")) {
            Intent scan = new Intent(this, BarcodeScan.class);
            startActivityForResult(scan, mRowID);
        } else {
            DialogInterface.OnClickListener dialogClickListener = new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    switch (which) {
                        case DialogInterface.BUTTON_NEGATIVE:
                            return;

                        case DialogInterface.BUTTON_POSITIVE:
                            mClueData[mRowID][1] = "";
                            setSingleTextAndChecks(mRowID);
                            writeClueData();
                            Intent scan = new Intent(ActivitySetClues.this, BarcodeScan.class);
                            startActivityForResult(scan, mRowID);
                    }
                }
            };
            AlertDialog.Builder builder = new AlertDialog.Builder(new ContextThemeWrapper(this, R.style.AlertDialogCustom));
            builder.setMessage(getResources().getString(R.string.setCluesRemoveCode, mClueData[mRowID][1])).setPositiveButton(getResources().getString(R.string.allYes), dialogClickListener)
                    .setNegativeButton(getResources().getString(R.string.allNo), dialogClickListener).show();
        }

    }

    public void goBackHome(View view) {
        onBackPressed();
    }

    public void completeSetUp(View view) {

        boolean mIsDataOk = true;

        if (mClueData[0][0].equals("") || mClueData[0][1].equals("")) {
            mIsDataOk = false;
        }

        for (int i = 1; i < 9; i++) {
            if ((mClueData[i][0].equals("") ^ mClueData[i][1].equals("")) || (mClueData[i][0].equals("") && !mClueData[i][0].equals(""))) {
                mIsDataOk = false;
            }
        }

        if (mIsDataOk) {
            DialogInterface.OnClickListener dialogClickListener = new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    switch (which) {
                        case DialogInterface.BUTTON_NEGATIVE:
                            return;
                        case DialogInterface.BUTTON_POSITIVE:
                            writeClueData();
                            writeHuntStart();
                            onBackPressed();
                    }
                }
            };
            AlertDialog.Builder builder = new AlertDialog.Builder(new ContextThemeWrapper(this, R.style.AlertDialogCustom));
            builder.setMessage(getResources().getString(R.string.setCluesBeginHunt));
            builder.setPositiveButton(getResources().getString(R.string.allYes), dialogClickListener);
            builder.setNegativeButton(getResources().getString(R.string.allNo), dialogClickListener);
            builder.show();
        } else {
            AlertDialog.Builder builder = new AlertDialog.Builder(new ContextThemeWrapper(this, R.style.AlertDialogCustom));
            builder.setMessage(getResources().getString(R.string.setCluesCantFinish));
            builder.setNeutralButton(getResources().getString(R.string.ok), null);
            builder.show();
        }
    }

    public void resetData(View view) {
        DialogInterface.OnClickListener dialogClickListener = new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                switch (which) {
                    case DialogInterface.BUTTON_NEGATIVE:
                        return;

                    case DialogInterface.BUTTON_POSITIVE:
                        eraseClueData();
                        setAllTextAndChecks();
                }
            }
        };
        AlertDialog.Builder builder = new AlertDialog.Builder(new ContextThemeWrapper(this, R.style.AlertDialogCustom));
        builder.setMessage(getResources().getString(R.string.setCluesResetAllWarning));
        builder.setPositiveButton(getResources().getString(R.string.allYes), dialogClickListener);
        builder.setNegativeButton(getResources().getString(R.string.allNo), dialogClickListener);
        builder.show();
    }

    public void returnTextInput(View view) {
        saveTextInput();
        setSingleTextAndChecks(mRowID);
        textInputInvisible();
    }

    public void clearInput(View view) {
        TextView mTextInput = findViewById(R.id.setCluesTextInput);
        mTextInput.setText("");
    }

    public void doneInput(View view) {
        saveTextInput();
        setSingleTextAndChecks(mRowID);
        textInputInvisible();
    }

    /* *****************
        Display Changes
       ***************** */
    private void setAllTextAndChecks() {

        for (int i = 0; i < mClueData.length; i++) {

            setSingleTextAndChecks(i);

        }

    }

    private void setSingleTextAndChecks(int mRowNumber) {
        int mTextID = getResources().getIdentifier("setClueText" + mRowNumber, "id", getPackageName());
        int mTextCheckID = getResources().getIdentifier("setClueTextCheck" + mRowNumber, "id", getPackageName());
        int mScanCheckID = getResources().getIdentifier("setClueCodeCheck" + mRowNumber, "id", getPackageName());

        TextView mText = findViewById(mTextID);
        ImageButton mTextCheck = findViewById(mTextCheckID);
        ImageButton mScanCheck = findViewById(mScanCheckID);

        int mHasText = mClueData[mRowNumber][0].equals("") ? 0 : 1;
        int mHasCode = mClueData[mRowNumber][1].equals("") ? 0 : 1;

        mTextCheck.setAlpha(mCheckStatus[mHasText]);
        mTextCheck.setImageResource(mClueChecks[mHasText]);

        mScanCheck.setAlpha(mCheckStatus[mHasCode]);
        mScanCheck.setImageResource(mCodeChecks[mHasCode]);

        if (mClueData[mRowNumber][0].equals("")) {
            mText.setText(getResources().getString(R.string.setCluesNotSet));
        } else {
            mText.setText(mClueData[mRowNumber][0]);
        }
    }

    private void textInputVisible(int mRow) {
        EditText mTextInput = findViewById(R.id.setCluesTextInput);

        mTextInput.setText(mClueData[mRow][0]);
        mTextInput.requestFocus();
        mTextInput.setSelection(mTextInput.getText().length());

        mTextInput.setVisibility(View.VISIBLE);

        findViewById(R.id.SetCluesTextInputOverlay).setVisibility(View.VISIBLE);
        findViewById(R.id.setCluesInputClear).setVisibility(View.VISIBLE);
        findViewById(R.id.setCluesInputDone).setVisibility(View.VISIBLE);

        mInputMethod = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        mInputMethod.showSoftInput(mTextInput, InputMethodManager.SHOW_IMPLICIT);
    }

    private void textInputInvisible() {
        TextView mTextInput = findViewById(R.id.setCluesTextInput);

        mTextInput.setVisibility(View.INVISIBLE);

        findViewById(R.id.SetCluesTextInputOverlay).setVisibility(View.INVISIBLE);
        findViewById(R.id.setCluesInputClear).setVisibility(View.INVISIBLE);
        findViewById(R.id.setCluesInputDone).setVisibility(View.INVISIBLE);

        mInputMethod = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        mInputMethod.hideSoftInputFromWindow(mTextInput.getWindowToken(), 0);
    }

    /* *********
        Methods
       ********* */
    private boolean codeAlreadyAssigned(String mCode) {

        boolean mAssigned = false;

        for (String[] codes : mClueData) {
            if (codes[1].equals(mCode)) {
                mAssigned = true;
                break;
            }
        }

        return mAssigned;

    }

    /* *****************
        Data Read/Write
       ***************** */
    private void loadClueData() {
        mDataStore = getSharedPreferences(getResources().getString(R.string.dataClueData), Context.MODE_PRIVATE);
        for (int i = 0; i < 10; i++) {
            mClueData[i][0] = mDataStore.getString("clue" + i, "");
            mClueData[i][1] = mDataStore.getString("code" + i, "");
        }
    }

    private void writeClueData() {
        mDataStoreEditor = mDataStore.edit();
        for (int i = 0; i < 10; i++) {
            mDataStoreEditor.putString("clue" + i, mClueData[i][0]);
            mDataStoreEditor.putString("code" + i, mClueData[i][1]);
        }
        mDataStoreEditor.apply();
    }

    private void writeHuntStart() {
        mDataStoreEditor = mDataStore.edit();
        mDataStoreEditor.putBoolean(getResources().getString(R.string.dataHuntStarted), true);
        mDataStoreEditor.commit();
    }

    private void eraseClueData() {
        mDataStoreEditor = mDataStore.edit();
        for (int i = 0; i < 10; i++) {
            mClueData[i][0] = "";
            mClueData[i][1] = "";
            mDataStoreEditor.putString("clue" + i, "");
            mDataStoreEditor.putString("code" + i, "");
        }
        mDataStoreEditor.apply();
    }

    private void saveTextInput() {
        TextView mTextInput = findViewById(R.id.setCluesTextInput);
        mClueData[mRowID][0] = mTextInput.getText().toString();
    }
}

File: app/src/main/java/com/xn1ch1/qrscavengerhunt/ActivitySetClues.kt
package com.xn1ch1.qrscavengerhunt

import android.Manifest
import android.annotation.SuppressLint
import android.app.AlertDialog
import android.content.DialogInterface
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.provider.Settings
import android.view.LayoutInflater
import android.view.View
import android.widget.*
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat

class ActivitySetClues : BaseActivity() {

	private lateinit var clueTable: LinearLayout

	private var clues = HashMap<Int, Clue>(10)

	private lateinit var cluePrefs: SharedPreferences

	private var clueSet = 0

	private lateinit var clueInput: EditText
	private lateinit var clueInputDone: TextView
	private lateinit var clueInputClear: TextView
	private lateinit var clueInputCancel: TextView
	private lateinit var clueInputOverlay: ConstraintLayout

	@SuppressLint("InflateParams")
	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_set_clues)

		cluePrefs = this.getSharedPreferences(Prefs.PREFS_FILENAME, 0)

		clueTable = findViewById(R.id.clueTable)
		clueInput = findViewById(R.id.setCluesTextInput)
		clueInputClear = findViewById(R.id.setCluesInputClear)
		clueInputCancel = findViewById(R.id.setCluesInputCancel)
		clueInputDone = findViewById(R.id.setCluesInputDone)
		clueInputOverlay = findViewById(R.id.clueInputOverlay)

		clueInputClear.setOnClickListener {
			clueInput.setText("")
			clueInput.requestFocus()
		}
		clueInputCancel.setOnClickListener {
			hideClueInput()
		}
		clueInputDone.setOnClickListener {
			with (clues[clueSet]!!) {
				clueText = clueInput.text.toString()
				hideClueInput()
				setClueCheck()
				setClueText()
			}
		}

		for (i in 0..9) {

			clues[i] = Clue()

			with (clues[i]!!) {

				number = i

				row = LayoutInflater.from(applicationContext).inflate(
						R.layout.set_clue_row,
						null
				) as ConstraintLayout

				rowMessage = row.findViewById(R.id.text)
				rowNumber = row.findViewById(R.id.number)
				rowClueButton = row.findViewById(R.id.textCheck)
				rowScanButton = row.findViewById(R.id.codeCheck)

				with (rowScanButton) {
					setOnClickListener {
						clueSet = i
						if (code.isNullOrEmpty()) {
							setCluesScanBarcode()
						} else {
							replaceCode()
						}
					}
				}

				with (rowClueButton) {
					setOnClickListener {
						clueSet = i
						clueInput.setText(clueText)
						showClueInput()
					}
				}

				rowNumber.text = (number + 1).toString()
				rowMessage.text = clueText
				clueTable.addView(row)

				setCodeCheck()
				setClueCheck()
				setClueText()

			}
		}

		findViewById<ImageView>(R.id.setCluesBackButton).setOnClickListener {
			onBackPressed()
		}
		findViewById<ImageView>(R.id.setCluesResetButton).setOnClickListener {
			resetData()
		}
		findViewById<ImageView>(R.id.setCluesFinishButton).setOnClickListener {
			startHunt()
		}

		hideClueInput()

	}

	override fun onBackPressed() {

		if (clueInputOverlay.visibility == View.VISIBLE) {
			hideClueInput()
		} else {
			super.onBackPressed()
			setResult(RESULT_OK, intent)
			finish()
		}

	}

	override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {

		if (resultCode == RESULT_OK) {
			if (requestCode == CAMERA_PERMISSION_REQUEST_CODE) {
				if (ContextCompat.checkSelfPermission(
								this,
								Manifest.permission.CAMERA
						) == PackageManager.PERMISSION_GRANTED
				) {
					openCameraWithScanner()
				}
			} else if (requestCode == ACT_BARCODE) {
				val code = data?.getStringExtra("barcode").toString()

				if (!code.isBlank() && codeExists(code)) {
					val builder = AlertDialog.Builder(this)
					with(builder) {
						setMessage(resources.getString(R.string.alreadyAssigned, code))
						setNeutralButton(resources.getString(R.string.ok), null)
						show()
					}
				} else {
					clues[clueSet]?.code = code
					clues[clueSet]?.setCodeCheck()
				}
			}
		}

	}

	private fun replaceCode() {

		val builder = AlertDialog.Builder(this)
		with (builder) {
			setMessage(getString(R.string.removeCode, clues[clueSet]?.code))
			setNegativeButton(getString(R.string.no), null)
			setPositiveButton(getString(R.string.yes)) { _, _ ->
				clues[clueSet]?.code = ""
				clues[clueSet]?.setClueCheck()
				setCluesScanBarcode()
			}
			show()
		}
	}

	private fun setCluesScanBarcode() {

		if (ContextCompat.checkSelfPermission(
						this,
						Manifest.permission.CAMERA
				) == PackageManager.PERMISSION_GRANTED
		) {
			openCameraWithScanner()
		} else {
			ActivityCompat.requestPermissions(
					this,
					arrayOf(Manifest.permission.CAMERA),
					CAMERA_PERMISSION_REQUEST_CODE
			)
		}

	}

	private fun showClueInput() {

		clueInputOverlay.visibility = View.VISIBLE
		clueInput.requestFocus()

		showKeyboard()

	}

	private fun hideClueInput() {

		hideKeyboard()

		clueInputOverlay.visibility = View.INVISIBLE
		clueInput.clearFocus()
		clueInput.setText("")


	}

	override fun onRequestPermissionsResult(
			requestCode: Int,
			permissions: Array<out String>,
			grantResults: IntArray
	) {
		super.onRequestPermissionsResult(requestCode, permissions, grantResults)

		if (requestCode == CAMERA_PERMISSION_REQUEST_CODE && grantResults.isNotEmpty()) {
			if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
				openCameraWithScanner()
			} else if (!ActivityCompat.shouldShowRequestPermissionRationale(
							this,
							Manifest.permission.CAMERA
					)
			) {
				val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
				val uri: Uri = Uri.fromParts("package", packageName, null)
				intent.data = uri
				startActivityForResult(intent, CAMERA_PERMISSION_REQUEST_CODE)
			}
		}

	}

	private fun openCameraWithScanner() {

		val intent = Intent(this, ActivityScanCode::class.java)
		startActivityForResult(intent, ACT_BARCODE)

	}

	private fun codeExists(code: String): Boolean {

		var exists = false

		clues.forEach { (_, clue) ->
			if (clue.code == code) {
				exists = true
			}
		}

		return exists

	}

	private fun startHunt() {

		var last = false
		var i = 0

		clues.forEach { (_, clue) ->

			if ((clue.code == "" && clue.clueText != "") || (clue.code != "" && clue.clueText == "")) {
            	val builder = AlertDialog.Builder(this)
				with (builder) {
					setMessage(getString(R.string.clueError, (clue.number + 1).toString()))
					setNeutralButton(getString(R.string.ok), null)
					show()
				}
				return
			} else if (clue.code == "" && clue.clueText == "") {
				last = true
			} else if (last && (clue.code != "" || clue.clueText != "")) {
				val builder = AlertDialog.Builder(this)
				with (builder) {
					setMessage(getString(R.string.clueError, (clue.number + 1).toString()))
					setNeutralButton(getString(R.string.ok), null)
					show()
				}
				return
			} else {
				i++
			}

		}

		prefs.clueCount = i
		prefs.huntStarted = true
		prefs.currentClue = 0
		onBackPressed()

	}

	private fun resetData() {
		val dialogClickListener = DialogInterface.OnClickListener { _, which ->
			when (which) {
				DialogInterface.BUTTON_NEGATIVE -> return@OnClickListener
				DialogInterface.BUTTON_POSITIVE -> {

					clues.forEach { (_, clue) ->
						clue.code = ""
						clue.clueText = ""
						clue.setClueText()
						clue.setClueCheck()
						clue.setCodeCheck()
					}

				}
			}
		}
		val builder = AlertDialog.Builder(this)
		with (builder) {
			setMessage(resources.getString(R.string.resetCluesQuestion))
			setPositiveButton(resources.getString(R.string.yes), dialogClickListener)
			setNegativeButton(resources.getString(R.string.no), dialogClickListener)
			show()
		}
	}

	inner class Clue {

		var number: Int = 0

		lateinit var row: ConstraintLayout
		lateinit var rowMessage: TextView
		lateinit var rowNumber: TextView
		lateinit var rowClueButton: ImageView
		lateinit var rowScanButton: ImageView

		var clueText: String?
			get() = cluePrefs.getString("clueText$number", "")
			set(value) {
				cluePrefs.edit().putString("clueText$number", value).apply()
			}

		var code: String?
			get() = cluePrefs.getString("code$number", "")
			set(value) {
				cluePrefs.edit().putString("code$number", value).apply()
			}

		fun setClueText() {
			if (clueText.isNullOrBlank()) {
				rowMessage.text = getString(R.string.clueNotSet)
				rowMessage.alpha = 0.5f
			} else {
				rowMessage.text = clueText
				rowMessage.alpha = 1f
			}
		}

		fun setClueCheck() {
			if (clueText.isNullOrBlank()) {
				rowClueButton.setImageResource(R.drawable.set_clue_clue_text_unchecked)
			} else {
				rowClueButton.setImageResource(R.drawable.set_clue_clue_text_checked)
			}
		}

		fun setCodeCheck() {
			if (code.isNullOrBlank()) {
				rowScanButton.setImageResource(R.drawable.set_clue_scan_unchecked)
			} else {
				rowScanButton.setImageResource(R.drawable.set_clue_scan_checked)
			}
		}


	}

}